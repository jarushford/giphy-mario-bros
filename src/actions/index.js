export const setPlayers = (players) => ({
  type: 'SET_PLAYERS',
  players
})

export const clearPlayers = () => ({
  type: 'CLEAR_PLAYERS'
})

export const setError = (error) => ({
  type: 'SET_ERROR',
  error
})

export const clearError = () => ({
  type: 'CLEAR_ERROR'
})