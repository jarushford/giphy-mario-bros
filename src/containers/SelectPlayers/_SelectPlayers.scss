.players {
  width: 50%;
  margin: 0px auto;
  padding: 20px 30px;
  background: $bg;
  border-radius: 15px;
  max-width: 500px;
  animation: fade .5s ease forwards;
}

.three-btn,
.four-btn {
  @include btn();
  margin: 30px auto;
}