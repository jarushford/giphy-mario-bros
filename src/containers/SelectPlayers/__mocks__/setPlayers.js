export const setPlayersThunk = jest.fn()