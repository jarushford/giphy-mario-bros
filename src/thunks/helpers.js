import { gifs } from '../utils/gifs'

export const getIDs = (players) => {
  const numNeeded = players * 5
  const IDs = []
  while (IDs.length !== numNeeded) {
    let randIndex = Math.floor(Math.random() * gifs.length)
    if (!IDs.includes(gifs[randIndex])) {
      IDs.push(gifs[randIndex])
    }
  }
  return IDs
}

export const buildPlayer = (players, response) => {
  let readyPlayers

  if (players === 3) {
    readyPlayers = [
      {
        player: 1,
        score: 0,
        gifs: response.splice(0, 5)
      },
      {
        player: 2,
        score: 0,
        gifs: response.splice(0, 5)
      },
      {
        player: 3,
        score: 0,
        gifs: response
      }
    ]
  } else {
    readyPlayers = [
      {
        player: 1,
        score: 0,
        gifs: response.splice(0, 5)
      },
      {
        player: 2,
        score: 0,
        gifs: response.splice(0, 5)
      },
      {
        player: 3,
        score: 0,
        gifs: response.splice(0, 5)
      },
      {
        player: 4,
        score: 0,
        gifs: response
      }
    ]
  }

  return readyPlayers
}