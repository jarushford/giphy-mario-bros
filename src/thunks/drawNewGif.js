import { apiKey } from '../utils/apiKey'
import { addNewGif, setError, setUnusedIDs } from '../actions'
import { getNewID } from './helpers'

export const drawNewGifThunk = (currentPlayer, unusedIDs) => {
  return async (dispatch) => {
    try {
      const newGif = getNewID(unusedIDs)
      const newUnusedIDs = unusedIDs.filter(ID => ID !== newGif)

      const url = `http://api.giphy.com/v1/gifs/${newGif}?api_key=${apiKey}`
      const response = await fetch(url)
      
      if (!response.ok) {
        throw Error('Could not fetch GIF')
      }

      const result = await response.json()
      dispatch(addNewGif({ gif: result.data, player: currentPlayer - 1}))
      dispatch(setUnusedIDs(newUnusedIDs))
    } catch (error) {
      dispatch(setError(error.message))
    }
  }
}