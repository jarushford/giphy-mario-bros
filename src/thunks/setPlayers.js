import { setPlayers, setError, setUnusedIDs } from '../actions'
import { getIDs, getUnusedIDs, buildPlayer, storeGifs } from './helpers'

export const setPlayersThunk = (players) => {
  return async (dispatch) => {
    const storage = JSON.parse(localStorage.getItem('gifs'))
    const IDs = getIDs(players)
    const unusedIDs = getUnusedIDs(IDs)

    if (!storage || !storage.length) {
      try{
        await storeGifs()
      } catch (error) {
        dispatch(setError(error.message))
      }
    }

    const gifs = JSON.parse(localStorage.getItem('gifs'))
    const response = gifs.filter(gif => IDs.includes(gif.id))
    
    const readyPlayers = buildPlayer(players, response)
    dispatch(setPlayers(readyPlayers))
    dispatch(setUnusedIDs(unusedIDs))
  }
}