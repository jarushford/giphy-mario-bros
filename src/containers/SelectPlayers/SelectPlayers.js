import React from 'react'
import { connect } from 'react-redux';
import { Link } from 'react-router-dom'
import { setPlayersThunk } from '../../thunks/setPlayers'
import { setCaption } from '../../actions'
import { captionHelper } from '../../utils/captions'
import PropTypes from 'prop-types'
import '../../main.scss'

export function SelectPlayers({ setPlayers, usedCaptions, setCaption }) {
  return (
    <section className="players">
      <h1 className="title">Giphy Mario Bros.</h1>
      <Link to='/home' onClick={() => {
        setPlayers(3)
        setCaption(captionHelper(usedCaptions))
      }}>
        <button className="three-btn">
          3 Players
        </button>
      </Link>
      <Link to='/home' onClick={() => {
        setPlayers(4)
        setCaption(captionHelper(usedCaptions))
      }}>
        <button className="four-btn">
          4 Players
        </button>
      </Link>
    </section>
  )
}

export const mapStateToProps = (state) => ({
  usedCaptions: state.caption
})

export const mapDispatchToProps = (dispatch) => ({
  setPlayers: (players) => dispatch(setPlayersThunk(players)),
  setCaption: (caption) => dispatch(setCaption(caption))
})

export default connect(mapStateToProps, mapDispatchToProps)(SelectPlayers)

SelectPlayers.propTypes = {
  usedCaptions: PropTypes.array.isRequired,
  setPlayers: PropTypes.func.isRequired,
  setCaption: PropTypes.func.isRequired
}