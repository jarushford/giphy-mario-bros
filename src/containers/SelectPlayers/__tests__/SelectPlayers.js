import React from 'react'
import { SelectPlayers, mapStateToProps, mapDispatchToProps } from '../SelectPlayers'
import { shallow } from 'enzyme'
import { setCaption } from '../../../actions'
import { setPlayersThunk } from '../../../thunks/setPlayers'

jest.mock('../../../thunks/setPlayers')

describe('SelectPlayers', () => {
  let mockSetPlayers
  let mockSetCaption
  let mockUsedCaptions
  let wrapper

  beforeEach(() => {
    mockSetPlayers = jest.fn()
    mockSetCaption = jest.fn()
    mockUsedCaptions = ['Test1', 'Test2']
    wrapper = shallow(<SelectPlayers 
        setPlayers={mockSetPlayers}
        setCaption={mockSetCaption}
        usedCaptions={mockUsedCaptions}
      />)
  })

  describe('SelectPlayers Component', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call setPlayers on click for three players', () => {
      wrapper.find('Link').first().simulate('click')

      expect(mockSetPlayers).toBeCalledWith(3)
    })

    it('should call setCaption on click for three players', () => {
      wrapper.find('Link').first().simulate('click')

      expect(mockSetCaption).toBeCalled()
    })

    it('should call setPlayers on click for four players', () => {
      wrapper.find('Link').last().simulate('click')

      expect(mockSetPlayers).toBeCalledWith(4)
    })

    it('should call setCaption on click for four players', () => {
      wrapper.find('Link').last().simulate('click')

      expect(mockSetCaption).toBeCalled()
    })
  })

  describe('mapStateToProps', () => {
    it('should return a props object with usedCaptions', () => {
      const mockState = {
        caption: mockUsedCaptions,
        judge: 1,
        error: ''
      }
      const expected = {
        usedCaptions: mockUsedCaptions
      }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    let mockDispatch

    beforeEach(() => {
      mockDispatch = jest.fn()
    })
    
    it('should dispatch setCaption when the prop setPlayers is called', () => {
      const expected = setCaption()

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setCaption()

      expect(mockDispatch).toBeCalledWith(expected)
    })

    it('should dispatch setPlayers when the prop setPlayers is called', () => {
      const expected = setPlayersThunk()

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setPlayers()

      expect(mockDispatch).toBeCalledWith(expected)
    })
  })
})