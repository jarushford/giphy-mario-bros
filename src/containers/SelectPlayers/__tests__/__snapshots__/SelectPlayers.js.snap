// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectPlayers SelectPlayers Component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectPlayers
    setCaption={[MockFunction]}
    setPlayers={[MockFunction]}
    usedCaptions={
      Array [
        "Test1",
        "Test2",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="title"
        >
          Giphy Mario Bros.
        </h1>,
        <Link
          onClick={[Function]}
          replace={false}
          to="/home"
        >
          <button
            className="three-btn"
          >
            3 Players
          </button>
        </Link>,
        <Link
          onClick={[Function]}
          replace={false}
          to="/home"
        >
          <button
            className="four-btn"
          >
            4 Players
          </button>
        </Link>,
      ],
      "className": "players",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Giphy Mario Bros.",
          "className": "title",
        },
        "ref": null,
        "rendered": "Giphy Mario Bros.",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <button
            className="three-btn"
          >
            3 Players
          </button>,
          "onClick": [Function],
          "replace": false,
          "to": "/home",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "3 Players",
            "className": "three-btn",
          },
          "ref": null,
          "rendered": "3 Players",
          "type": "button",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <button
            className="four-btn"
          >
            4 Players
          </button>,
          "onClick": [Function],
          "replace": false,
          "to": "/home",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "4 Players",
            "className": "four-btn",
          },
          "ref": null,
          "rendered": "4 Players",
          "type": "button",
        },
        "type": [Function],
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="title"
          >
            Giphy Mario Bros.
          </h1>,
          <Link
            onClick={[Function]}
            replace={false}
            to="/home"
          >
            <button
              className="three-btn"
            >
              3 Players
            </button>
          </Link>,
          <Link
            onClick={[Function]}
            replace={false}
            to="/home"
          >
            <button
              className="four-btn"
            >
              4 Players
            </button>
          </Link>,
        ],
        "className": "players",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Giphy Mario Bros.",
            "className": "title",
          },
          "ref": null,
          "rendered": "Giphy Mario Bros.",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <button
              className="three-btn"
            >
              3 Players
            </button>,
            "onClick": [Function],
            "replace": false,
            "to": "/home",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "3 Players",
              "className": "three-btn",
            },
            "ref": null,
            "rendered": "3 Players",
            "type": "button",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <button
              className="four-btn"
            >
              4 Players
            </button>,
            "onClick": [Function],
            "replace": false,
            "to": "/home",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "4 Players",
              "className": "four-btn",
            },
            "ref": null,
            "rendered": "4 Players",
            "type": "button",
          },
          "type": [Function],
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
